### ISSUE TITLE
create bash tutorials

### ISSUE THREAD

**Issue by patrickmalsom** [practical-scientific-computing/practical-scientific-computing.github.io#8]
# Shell list of topics

Please add/edit this list if there are other topics worth covering. I would like this to be a useful resource for researchers even after this workshop is over, so if there are other commands you use on a semi-weekly basis **please add them!**
## Tutorial topics
- [x] getting around: ls cd mkdir pwd rm mv cp ...
- [x] UNIX directory tree
- [x] file permissions
- [ ] sudo
- [ ] pipes
- [ ] redirection
- [ ] job control: ctrl-z, ctrl-c, bg, fg, disown, &
- [x] ssh, scp
- [x] tmux, screen
- [ ] history, ctrl-r
- [x] man
- [ ] exit codes
- [x] bash scripts
- [x] whereis, which
- [ ] time/date
- [ ] sub shells (dollarsign and backtick)
- [x] editing files
- [x] grep
- [x] cat
- [x] tail
- [x] find
## Adv topics
- [ ] chaining multiple commands (;, &&, ||)
- [ ] redirecting to file descriptors (e.g., 2>&1)
- [ ] read file line by line (while read; blah; done < file)
- [ ] process states (S, R, Z, D, etc...)
- [ ] pushd, popd, cd -
- [ ] list expansion (e.g., mkdir {1,2,3,4})
- [ ] `termplot` alias with gnuplot 
- [ ] downloading files with curl and wget
- [ ] gzip/tar

**Comment by mpbelhorn** [comment]
I've added some topics (UNIX directory tree, sudo, interrupt) and am working on a shell basics tutorial that so far covers:
various shells, prompt, command syntax, whoami, pwd, hostname, ls, cd, mv, cp, rm, touch, mkdir, rmdir, directory tree

**Comment by patrickmalsom** [comment]
Possible breakdown of topics for tutorials. Please state which tutorial you are writing here so we aren't stepping on each other's toes when writing. Feel free to edit/add/remove items from the list. This is only a stab at organizing things, and I really have no idea what I am doing...

Thanks to @LincolnBryant for helping get this list together.

@mpbelhorn can you include man pages as well in your intro page?

# Tutorial topics

## Basic shell commands
- getting around: ls cd mkdir pwd rm mv cp ...
- echo
- UNIX directory tree
- possibly cover a bit more about '.' '..' '~' '-' and other special ones
- man pages
- file permissions

## Editing and manipulating text files
- editing files
  - nano
  - vim
  - emacs
- grep
- cat
- head/tail
- less

## Pipelining, redirection, and job control
- ;
- |
- >
- <
- &&
- ||
- Ctrl-Z
- Ctrl-C
- bg
- fg
- disown
- &

## time savers
- history
- ctrl-r
- ctrl-a/ctrl-e
- alias
- find
- !! 

## Remote access
- ssh
- scp
- rsync
- screen

## Shell scripts
- bash scripts (#!/bin/sh -x)
- sub shells (dollarsign and backtick)
- loops and conditionals
- exit codes
- time/date

# Adv topics
- chaining multiple commands (;, &&, ||)
- redirecting to file descriptors (e.g., 2>&1)
- read file line by line (while read; blah; done < file)
- process states (S, R, Z, D, etc...)
- pushd, popd <-- cool wizard spells but not really useful for 95% of cases
- list expansion (e.g., mkdir {1,2,3,4})
- downloading files with curl and wget
- gzip/tar
- administration (sudo)

## Examples
- `termplot` alias with gnuplot (make this an example of aliasing making life easier)

# Reference commands
- just a file that covers many of the above commands in a list (something like below)
  - https://swc-osg-workshop.github.io/2014-12-15-UChicago/novice/ref/01-shell.html

**Comment by mpbelhorn** [comment]
I'm enhancing the first tutorial page now with your suggestions and it occurs to me we need to discuss environment variables somewhere, especially if folks are working on remote systems where the environment can be crazy. Folks will likely need to know about at least $PATH, $HOME, $EDITOR, $TERM, $PAGER, $DISPLAY, and $PYTHONPATH

**Comment by adamdddave** [comment]
I'll go ahead and take remote access and shell scripts

**Comment by mpbelhorn** [comment]
Is this really closed? Do we have someone working on everything in the list?

**Comment by LincolnBryant** [comment]
I am writing some docs on pipelining, redirection, and job control.

**Comment by mpbelhorn** [comment]
I can take **editing and manipulating text files**. Although I have no intention of going into detail on editors - I'll just give them enough information to be dangerous and point 'em in the direction to learn those on their own.

I'll also throw some 'find', 'sed', and 'regexp' recipes in with that tutorial.

**Comment by LincolnBryant** [comment]
Pat and I talked about this at one point, but it would be really nice to have some supplementary material that solely covers regular expressions. They're incredibly useful for any question that starts with "How do I find all _foo_ for condition _bar_?". 

Just knowing to tack on "regexp" to a search query goes a long way.

**Comment by adamdddave** [comment]
I'm doing remote stuff (ssh, scp, rsync, etc) and shell scripts. I
anticipate starting this stuff at the end of the first lecture or the
beginning of the second.

I agree, regexp is extremely useful and could take up a whole lecture.

**Comment by mpbelhorn** [comment]
I agree completely. Full coverage of regexp would would be super useful,
but I don't think I can have that written by Friday. I'd be glad to work on
it though. I have a condensed cheat sheet that I've been working on for
years that would make a good basis for a regexp tutorial.

**Comment by patrickmalsom** [comment]
Ok so we are going to present this stuff on Friday so making some sort of plan is probably in order. 

I am going to start the session by making sure everyone has access to the remote GNU+Linux machine that is set up for this class. This will require ssh, but I think I will do this without a tutorial and save the remote stuff for later in the meeting.

@mpbelhorn, I am expecting that you will want to be teaching for most of this meeting as you have written the entire first page. Please comment if this is not your plan.

### First Meeting
1. Basic shell commands
2. Editing files
3. Remote access 

### Second meeting
1. Pipelining
2. Job control
3. Time savers
4. ...

I also think a regexp page is a good idea but will most likely be in the advanced tutorials, (hopefully linked to from tutorial pages). Basic regexp syntax will obviously be needed for any sed discussion.

@mpbelhorn: I think a cheat sheet for shell commands will be **very useful** if you have one drafted up. Do you want to work on that and I will take the editor section? I am fine with you writing both if you are up for it. Let me know what you want to do.

**Comment by adamdddave** [comment]
I think this is a sensible plan.

**Comment by mpbelhorn** [comment]
Pat, I was planning on presenting on Friday. I can write up a cheat sheet
of commands if you want to write about editors. I'm on my phone walking to
something now, but I'll get in touch later tonight about what do.

**Comment by adamdddave** [comment]
I'm covering tmux and screen in the remote tutorial, fyi

**Comment by patrickmalsom** [comment]
I would personally only introduce screen and avoid tmux. Tmux is not installed on many of the scientific clusters but screen will be. 

Just my 2 cents...

**Comment by adamdddave** [comment]
Done
